kurtosis-devnet/pkg/build/docker.go

```diff
@@ -173,63 +227,110 @@ type templateData struct {
 	ProjectName string
 }
 
-// Build executes the docker build command for the given project and image tag
-// Note: the returned image tag is the image ID, so we don't accidentally
-// de-duplicate steps that should not be de-duplicated.
+// Build ensures the docker image for the given project is built, respecting concurrency limits.
+// It blocks until the specific requested build is complete. Other builds may run concurrently.
 func (b *DockerBuilder) Build(projectName, imageTag string) (string, error) {
-	if builtImage, ok := b.builtImages[projectName]; ok {
-		return builtImage, nil
+	b.mu.Lock()
+	state, exists := b.buildStates[projectName]
+	if !exists {
+		state = &buildState{
+			done: make(chan struct{}),
+		}
+		b.buildStates[projectName] = state
+		b.mu.Unlock()
+
+		state.once.Do(func() {
+			err := b.executeBuild(projectName, imageTag, state)
+			if err != nil {
+				state.err = err
+				state.result = ""
+			}
+			close(state.done)
+		})
+	} else {
+		b.mu.Unlock()
+		<-state.done
 	}
```

Comment:
We could only lock when needed.
I wonder if this would read nicer (and perhaps run a bit faster as well):

```
state, exists := b.buildStates[projectName]
if !exists {
  b.mu.Lock()
  // do all the things
  b.mu.Unlock()
} else {
  <-state.done
}
```

Just an idea.

Reply:
Good idea.